{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def read_fasta_file(filename):\n",
    "    \"\"\"\n",
    "    Reads the given FASTA file f and returns a dictionary of sequences.\n",
    "\n",
    "    Lines starting with ';' in the FASTA file are ignored.\n",
    "    \"\"\"\n",
    "    sequences_lines = {}\n",
    "    current_sequence_lines = None\n",
    "    with open(filename) as fp:\n",
    "        for line in fp:\n",
    "            line = line.strip()\n",
    "            if line.startswith(';') or not line:\n",
    "                continue\n",
    "            if line.startswith('>'):\n",
    "                sequence_name = line.lstrip('>')\n",
    "                current_sequence_lines = []\n",
    "                sequences_lines[sequence_name] = current_sequence_lines\n",
    "            else:\n",
    "                if current_sequence_lines is not None:\n",
    "                    current_sequence_lines.append(line)\n",
    "    sequences = {}\n",
    "    for name, lines in sequences_lines.items():\n",
    "        sequences[name] = ''.join(lines)\n",
    "    return sequences\n",
    "\n",
    "def create_fasta_dict(n):\n",
    "    d = {'genome' : [], 'annotation' : [], 'genome_name' : []}\n",
    "    for i in n:\n",
    "        d['genome_name'].append('genome' + str(i))\n",
    "        d['genome'].append(read_fasta_file('genome' + str(i) + '.fa')['genome' + str(i)])\n",
    "        if i < 6:\n",
    "            d['annotation'].append(read_fasta_file('annotation'+ str(i) + '.fa')['annotation' + str(i)])\n",
    "        else:\n",
    "            d['annotation'].append(None)\n",
    "    return d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\n"
     ]
    }
   ],
   "source": [
    "train_set = create_fasta_dict([1,2,3,4,5])\n",
    "test_set = create_fasta_dict([6,7,8,9,10])\n",
    "print(len(train_set['genome']))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "def viterbi(transition, emission, pi, hidden, sequence, observables):\n",
    "    #init viterbi table\n",
    "    viterbi_table = np.zeros((len(hidden), len(sequence)))\n",
    "    viterbi_table = viterbi_table.astype(np.float)\n",
    "    #init result list Z\n",
    "    Z = [0] * len(sequence)\n",
    "    \n",
    "    for i in range(len(hidden)):\n",
    "        viterbi_table[i,0] = np.log(pi[i]) + np.log(emission[i, observables[sequence[0]]])\n",
    "    for n in range(1,len(sequence)):\n",
    "        if n%100000==0:\n",
    "            print(n)\n",
    "        for k in range(len(hidden)):\n",
    "            viterbi_table[k,n] = float(\"-inf\")\n",
    "            if emission[k, observables[sequence[n]]] != float(0):\n",
    "                for j in range(len(hidden)):\n",
    "                    if transition[j,k] != float(0):\n",
    "                        viterbi_table[k,n] = max(viterbi_table[k,n], np.log(emission[k, observables[sequence[n]]]) + viterbi_table[j, n - 1] + np.log(transition[j,k]))\n",
    "    Z[-1] = np.argmax(viterbi_table[:,-1])\n",
    "    print(\"backtracking\")\n",
    "    for n in reversed(range(len(sequence)-1)):\n",
    "        if n%100000==0:\n",
    "            print(n)\n",
    "        state = float(\"-inf\")\n",
    "        for k in range(len(hidden)):\n",
    "            z = np.log(emission[Z[n+1], observables[sequence[n+1]]]) + viterbi_table[k,n] + np.log(transition[k, Z[n+1]])\n",
    "            if z > state:\n",
    "                state = z\n",
    "                index = k\n",
    "        Z[n] = index\n",
    "    return Z\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#training by counting\n",
    "SYMBOL_DICT={'A':0,'C':1,'G':2,'T':3}\n",
    "NUMBER_OF_SYMBOLS=4\n",
    "numberOfStates=43\n",
    "A=np.zeros([numberOfStates,numberOfStates])\n",
    "startVector=np.zeros([numberOfStates])\n",
    "startVector[0]=1.0\n",
    "emissionTabel=np.zeros([numberOfStates,NUMBER_OF_SYMBOLS])\n",
    "\n",
    "def updateEmissionTabel(observations,states):\n",
    "    for i in range(len(states)):\n",
    "        emissionTabel[states[i],SYMBOL_DICT[observations[i]]]+=1\n",
    "    return(emissionTabel)\n",
    "def counting(genome,annotation, A = None, emissionTabel = None):\n",
    "    if A == None:\n",
    "        A=np.zeros([numberOfStates,numberOfStates])\n",
    "        emissionTabel = np.zeros([numberOfStates,NUMBER_OF_SYMBOLS])\n",
    "    a = genome\n",
    "    b = annotation\n",
    "    j=0\n",
    "    jo=-1\n",
    "    i = 1\n",
    "    state=0\n",
    "    while j <len(a):\n",
    "        if j==jo:\n",
    "            print(\"her\", state,jo,j,b[j:j+4], a[j:j+4])\n",
    "            break\n",
    "        if state==0:\n",
    "            if b[j]=='C':\n",
    "                if a[j:j+3]==\"TTG\": #state=1\n",
    "                    A[0,1]=A[0,1]+1\n",
    "                    state=3\n",
    "                    j+=3\n",
    "                elif a[j:j+3]==\"GTG\": #state=4\n",
    "                    A[0,4]+=1\n",
    "                    state=6\n",
    "                    j+=3\n",
    "                elif a[j:j+3]==\"ATG\": #state=7\n",
    "                    A[0,7]+=1\n",
    "                    state=9\n",
    "                    j+=3\n",
    "                else:\n",
    "                    A[0,0]+=1\n",
    "                    emissionTabel[0,SYMBOL_DICT[a[j]]]+=1\n",
    "                    j+=1\n",
    "            elif b[j]=='R':\n",
    "                if a[j:j+3]==\"CTA\": #state=22\n",
    "                    A[0,22]+=1\n",
    "                    state=24\n",
    "                    j+=3\n",
    "                elif a[j:j+3]==\"TTA\": #state=25\n",
    "                    A[0,25]+=1\n",
    "                    state=27\n",
    "                    j+=3\n",
    "                elif a[j:j+3]==\"TCA\": #state=28\n",
    "                    A[0,28]+=1\n",
    "                    state=30\n",
    "                    j+=3\n",
    "                else:\n",
    "                    A[0,0]+=1\n",
    "                    emissionTabel[0,SYMBOL_DICT[a[j]]]+=1\n",
    "                    j+=1\n",
    "                    \n",
    "            else:\n",
    "                A[0,0]+=1\n",
    "                emissionTabel[0,SYMBOL_DICT[a[j]]]+=1\n",
    "                j+=1\n",
    "        elif (state>=10) & (state<=12):\n",
    "            if b[j:j+4]==\"CCCN\":\n",
    "                if state!=12:\n",
    "                    print(\"UPS 4\",i,j,b[j:j+4],a[j:j+4])\n",
    "                else:\n",
    "                    if a[j:j+3]==\"TAG\": #state=13\n",
    "                        A[12,13]+=1\n",
    "                        state=15\n",
    "                        j+=3\n",
    "                    elif a[j:j+3]==\"TGA\": #state=16\n",
    "                        A[12,16]+=1\n",
    "                        state=18\n",
    "                        j+=3\n",
    "                    elif a[j:j+3]==\"TAA\": #state=19\n",
    "                        A[12,19]+=1\n",
    "                        state=21\n",
    "                        j+=3\n",
    "                    else:\n",
    "                        print(\"UPS 3\",i,j,b[j:j+4],a[j:j+4])\n",
    "                        state=0\n",
    "            elif b[j:j+4]==\"CCCC\":\n",
    "                emissionTabel = updateEmissionTabel(a[j:j+3],[10,11,12])\n",
    "                A[12,10]+=1\n",
    "                state=12\n",
    "                j+=3\n",
    "            else:\n",
    "                print(\"UPS 1\",i,j,b[j:j+4],a[j:j+4])\n",
    "                state = 0\n",
    "                for char in b[j:j+4]:\n",
    "                    if char == 'C':\n",
    "                        j += 1\n",
    "\n",
    "        elif (state>=31) & (state<=33):\n",
    "            if b[j:j+4]==\"RRRN\":\n",
    "                if state!=33:\n",
    "                    print(\"UPS 5\",i,j,b[j:j+4],a[j:j+4])\n",
    "                else:\n",
    "                    if a[j:j+3]==\"CAC\": #state=34\n",
    "                        A[33,34]+=1\n",
    "                        state=36\n",
    "                        j+=3\n",
    "                    elif a[j:j+3]==\"CAT\": #state=37\n",
    "                        A[33,37]+=1\n",
    "                        state=39\n",
    "                        j+=3\n",
    "                    elif a[j:j+3]==\"CAA\": #state=40\n",
    "                        A[33,40]+=1\n",
    "                        state=42\n",
    "                        j+=3\n",
    "                    else:\n",
    "                        print(\"UPS 6\",i,j,b[j:j+4],a[j:j+4])\n",
    "                        state = 0\n",
    "                        j += 3\n",
    "            elif b[j:j+4]==\"RRRR\":\n",
    "                emissionTabel = updateEmissionTabel(a[j:j+3],[31,32,33])\n",
    "                A[33,31]+=1\n",
    "                state=33\n",
    "                j+=3\n",
    "            else:\n",
    "                print(\"UPS 2\",i,j,b[j:j+4],a[j:j+4])\n",
    "                state = 0\n",
    "                j += 3\n",
    "        elif (state==36) | (state==39)|(state==42)| (state==15) | (state==18)|(state==21):\n",
    "            if b[j]=='N':\n",
    "                A[state,0]+=1\n",
    "                emissionTabel[0,SYMBOL_DICT[a[j]]]+=1\n",
    "                state=0\n",
    "                j+=1\n",
    "            else:\n",
    "                print(\"UPS 7\",i,j,b[j],a[j]) \n",
    "        elif (state==3) | (state==6)|(state==9):\n",
    "            if b[j:j+4]==\"CCCC\":\n",
    "                emissionTabel = updateEmissionTabel(a[j:j+3],[10,11,12])\n",
    "                A[state,10] += 1\n",
    "                state=12\n",
    "                j+=3\n",
    "            else:\n",
    "                print(\"UPS 9\",i,j,b[j:j+4],a[j:j+4])\n",
    "        elif (state==24) | (state==27)|(state==30):\n",
    "            if b[j:j+4]==\"RRRR\":\n",
    "                emissionTabel = updateEmissionTabel(a[j:j+3],[31,32,33])\n",
    "                A[state, 31] += 1\n",
    "                state=33\n",
    "                j+=3\n",
    "            else:\n",
    "                print(\"UPS 10\",i,j,b[j:j+4],a[j:j+4])\n",
    "            \n",
    "        else:\n",
    "            print(\"UPS 8\",i,j,b[j],a[j])\n",
    "    \n",
    "    return(A, emissionTabel)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def set_fixed_parameters(A, emission):\n",
    "    emission[1,3] = 1\n",
    "    emission[2,3] = 1\n",
    "    emission[3,2] = 1\n",
    "    emission[4,2] = 1\n",
    "    emission[5,3] = 1\n",
    "    emission[6,2] = 1\n",
    "    emission[7,0] = 1\n",
    "    emission[8,3] = 1\n",
    "    emission[9,2] = 1\n",
    "    emission[13,3] = 1\n",
    "    emission[14,0] = 1\n",
    "    emission[15,2] = 1\n",
    "    emission[16,3] = 1\n",
    "    emission[17,2] = 1\n",
    "    emission[18,0] = 1\n",
    "    emission[19,3] = 1\n",
    "    emission[20,0] = 1\n",
    "    emission[21,0] = 1\n",
    "    emission[22,1] = 1\n",
    "    emission[23,3] = 1\n",
    "    emission[24,0] = 1\n",
    "    emission[25,3] = 1\n",
    "    emission[26,3] = 1\n",
    "    emission[27,0] = 1\n",
    "    emission[28,3] = 1\n",
    "    emission[29,1] = 1\n",
    "    emission[30,0] = 1\n",
    "    emission[34,1] = 1\n",
    "    emission[35,0] = 1\n",
    "    emission[36,1] = 1\n",
    "    emission[37,1] = 1\n",
    "    emission[38,0] = 1\n",
    "    emission[39,3] = 1\n",
    "    emission[40,1] = 1\n",
    "    emission[41,0] = 1\n",
    "    emission[42,0] = 1\n",
    "\n",
    "    A[1,2] = 1\n",
    "    A[2,3] = 1\n",
    "    A[3,10] = 1\n",
    "    A[4,5] = 1\n",
    "    A[5,6] = 1\n",
    "    A[6,10] = 1\n",
    "    A[7,8] = 1\n",
    "    A[8,9] = 1\n",
    "    A[9,10] = 1\n",
    "    A[10,11] = 1\n",
    "    A[11,12] = 1\n",
    "    A[19,20] = 1\n",
    "    A[13,14] = 1\n",
    "    A[14,15] = 1\n",
    "    A[15,0] = 1\n",
    "    A[16,17] = 1\n",
    "    A[17,18] = 1\n",
    "    A[18,0] = 1\n",
    "    A[20,21] = 1\n",
    "    A[21,0] = 1\n",
    "    A[22,23] = 1\n",
    "    A[23,24] = 1\n",
    "    A[24,31] = 1\n",
    "    A[25,26] = 1\n",
    "    A[26,27] = 1\n",
    "    A[27,31] = 1\n",
    "    A[28, 29] = 1\n",
    "    A[29,30] = 1\n",
    "    A[30,31] = 1\n",
    "    A[31,32] = 1\n",
    "    A[32,33] = 1\n",
    "    A[34,35] = 1\n",
    "    A[35,36] = 1\n",
    "    A[36,0] = 1\n",
    "    A[37,38] = 1\n",
    "    A[38,39] = 1\n",
    "    A[39,0] = 1\n",
    "    A[40,41] = 1\n",
    "    A[41,42] = 1\n",
    "    A[42,0] = 1\n",
    "    Atotal = A.sum(axis = 1, keepdims = True)\n",
    "    new_A = A / Atotal\n",
    "\n",
    "    emissiontotal = emission.sum(axis = 1, keepdims = True)\n",
    "    new_emission = emission / emissiontotal\n",
    "    return new_A,new_emission"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def fivefold_cross_validation(train_set):\n",
    "    observables = {'A' : 0, 'C' : 1, 'G' : 2, 'T' : 3}\n",
    "    results = []\n",
    "    names = []\n",
    "    n = len(train_set['genome'])\n",
    "    for i in range(n):\n",
    "        print(i)\n",
    "        trainingdata = train_set['genome'][:]\n",
    "        print(len(trainingdata))\n",
    "        training_annotation = train_set['annotation'][:]\n",
    "        validationdata = trainingdata.pop(i)\n",
    "        validation_annotation = training_annotation.pop(i)\n",
    "        names.append(train_set['genome_name'][i])\n",
    "        A = None\n",
    "        emissionTabel = None\n",
    "        for j in range(n-1):\n",
    "            print(\"test\", j)\n",
    "            A, emissionTabel = counting(trainingdata[j], training_annotation[j], A, emissionTabel)\n",
    "        A, emissionTabel = set_fixed_parameters(A, emissionTabel)\n",
    "        pi = np.zeros(43)\n",
    "        pi[0] = 1\n",
    "        prediction = viterbi(A, emissionTabel, pi, dict(zip(range(0,43), range(0,43))), validationdata, observables)\n",
    "        results.append(prediction)\n",
    "    return results\n",
    "\n",
    "def convert_prediction(results):\n",
    "    for i in range(len(results)):\n",
    "        for j in range(len(results[i])):\n",
    "            if results[i][j] == 0:\n",
    "                results[i][j] = 'N'\n",
    "            elif (results[i][j] >= 1) and (results[i][j] <= 21):\n",
    "                results[i][j] = 'C'\n",
    "            elif (results[i][j] >= 22) and (results[i][j] <= 42):\n",
    "                results[i][j] = 'R'\n",
    "    return results\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#test crossvalidation\n",
    "cv = fivefold_cross_validation(train_set)\n",
    "\n",
    "pred = convert_prediction(cv)\n",
    "for i in range(len(cv)):\n",
    "    with open(\"prediction%i.txt\"%i, \"w\") as file:\n",
    "        file.write(\"\".join(pred[i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "def predict_test_set(train_set,test_set):\n",
    "    observables = {'A' : 0, 'C' : 1, 'G' : 2, 'T' : 3}\n",
    "    results = []\n",
    "    names = []\n",
    "    n = len(train_set['genome'])\n",
    "    A = None\n",
    "    emissionTabel = None\n",
    "    trainingdata = train_set['genome'][:]\n",
    "    training_annotation = train_set['annotation'][:]\n",
    "    for i in range(n):\n",
    "        names.append(train_set['genome_name'][i])\n",
    "        A, emissionTabel = counting(trainingdata[i], training_annotation[i], A, emissionTabel)\n",
    "        print(A[0,0],emissionTabel[0,0])\n",
    "    A, emissionTabel = set_fixed_parameters(A, emissionTabel)\n",
    "    pi = np.zeros(43)\n",
    "    pi[0] = 1\n",
    "    testdata = test_set['genome'][:]\n",
    "    n=len(testdata)\n",
    "    for i in range(n):\n",
    "        prediction = viterbi(A, emissionTabel, pi, dict(zip(range(0,43), range(0,43))), testdata[i], observables)\n",
    "        results.append(prediction)\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#get predictions of test-set\n",
    "pt=predict_test_set(train_set,test_set)\n",
    "\n",
    "pred_test = convert_prediction(pt)\n",
    "for i in range(len(pt)):\n",
    "    t=i+6\n",
    "    with open(\"annotation%i.fa\"%t, \"w\") as file:\n",
    "        file.write(\">annotation\"+str(t)+\"\\n\")\n",
    "        for j in range(len(pred_test[i])):\n",
    "            file.write(pred_test[i][j])\n",
    "            if j%60==59:        \n",
    "                file.write('\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#get probabilities from A and emissionTabel\n",
    "def getMatrices():\n",
    "    A=None\n",
    "    emissionTabel=None\n",
    "    n = len(train_set['genome'])\n",
    "    trainingdata = train_set['genome'][:]\n",
    "    training_annotation = train_set['annotation'][:]\n",
    "    for i in range(n):\n",
    "        A, emissionTabel = counting(trainingdata[i], training_annotation[i], A, emissionTabel)\n",
    "        print(A[0,0],emissionTabel[0,0])\n",
    "    A, emissionTabel = set_fixed_parameters(A, emissionTabel)\n",
    "    return A,emissionTabel\n",
    "\n",
    "A,emissionTabel=getMatrices()\n",
    "with open(\"A.txt\", \"w\") as file:\n",
    "    for i in range(43):\n",
    "        for j in range(43):\n",
    "            if A[i,j]!=0:\n",
    "                \n",
    "                file.write(\"state \"+str(i)+\" state \"+str(j)+\" \"+str(A[i,j])+\"\\n\")\n",
    "symbolliste=['A','C','G','T']\n",
    "with open(\"emissionTabel.txt\", \"w\") as file:\n",
    "    for i in range(43):\n",
    "        for j in range(4):\n",
    "            if emissionTabel[i,j]!=0:\n",
    "                \n",
    "                file.write(\"state \"+str(i)+\" symbol \"+symbolliste[j]+\" \"+str(emissionTabel[i,j])+\"\\n\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pickle\n",
    "with open('pt.pickle', 'wb') as f:\n",
    "    pickle.dump(pt, f, pickle.HIGHEST_PROTOCOL)"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [Root]",
   "language": "python",
   "name": "Python [Root]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
